### PR TITLE
cudnnFindEx

### PR DESCRIPTION
**Pull request by aromnvidia**
Thank you guys again for your comments and recommendations on my prev. pull request.
Here is updated version of my implementation.

### CODE REVIEW

File: torch/lib/THC/THCGeneral.c [hunk 1]
```diff
@@ -871,6 +871,12 @@ void THCudaHostRecord(THCState *state, void *ptr)
 
 cudaError_t THCudaMemGetInfo(THCState *state,  size_t* freeBytes, size_t* totalBytes)
 {
+  size_t large = 0;
+  return THCudaMemGetInfo1(state, freeBytes, totalBytes, &large);
+}
+
+cudaError_t THCudaMemGetInfo1(THCState *state,  size_t* freeBytes, size_t* totalBytes, size_t* large)
```
Comment (apaszke) [inline torch/lib/THC/THCGeneral.c]: Can you rename it to `largestBlock`? It isn't obvious what `large` is

File: torch/lib/THC/THCGeneral.h.in [hunk 2]
```diff
@@ -209,6 +209,7 @@ THC_API void THCudaHostFree(THCState *state, void *ptr);
 THC_API void THCudaHostRecord(THCState *state, void *ptr);
 
 THC_API cudaError_t THCudaMemGetInfo(THCState *state, size_t* freeBytes, size_t* totalBytes);
+THC_API cudaError_t THCudaMemGetInfo1(THCState *state, size_t* freeBytes, size_t* totalBytes, size_t* large);
```
Comment (apaszke) [inline torch/lib/THC/THCGeneral.h.in]: Same here

File: torch/csrc/cudnn/Conv.cpp [hunk 3]
```diff
@@ -106,18 +106,63 @@ template<typename algo_t>
...
-        conv.wdesc.desc, conv.cdesc.desc, conv.odesc.desc, 1, &algoCount, &perfResults));
+    cudnnConvolutionFwdAlgo_t algo[] = {
+	    CUDNN_CONVOLUTION_FWD_ALGO_GEMM,
+	    CUDNN_CONVOLUTION_FWD_ALGO_FFT,
+	    CUDNN_CONVOLUTION_FWD_ALGO_FFT_TILING,
+	    CUDNN_CONVOLUTION_FWD_ALGO_IMPLICIT_GEMM,
+	    CUDNN_CONVOLUTION_FWD_ALGO_IMPLICIT_PRECOMP_GEMM,
+	    CUDNN_CONVOLUTION_FWD_ALGO_DIRECT,
+            CUDNN_CONVOLUTION_FWD_ALGO_WINOGRAD,
+            CUDNN_CONVOLUTION_FWD_ALGO_WINOGRAD_NONFUSED
```
Comment (apaszke) [inline torch/csrc/cudnn/Conv.cpp]: It seems that you have a mix of tabs and spaces in here

File: torch/lib/THC/THCGeneral.h.in (hunk 2 above)
Comment (apaszke) [inline torch/lib/THC/THCGeneral.h.in]: Maybe rename it to `THCudaMemGetInfoCached`?

File: torch/csrc/cudnn/Conv.cpp (hunk 3 above)
Comment (aromnvidia) [inline torch/csrc/cudnn/Conv.cpp]: Fixed

File: torch/lib/THC/THCGeneral.c (hunk 1 above)
Comment (aromnvidia) [inline torch/lib/THC/THCGeneral.c]: fixed

File: torch/lib/THC/THCGeneral.h.in (hunk 2 above)
Comment (aromnvidia) [inline torch/lib/THC/THCGeneral.h.in]: Fixed

File: torch/csrc/cudnn/Conv.cpp [hunk 4]
```diff
@@ -106,18 +106,63 @@ template<typename algo_t>
...
+         CUDNN_CONVOLUTION_FWD_ALGO_DIRECT,
+         CUDNN_CONVOLUTION_FWD_ALGO_WINOGRAD,
+         CUDNN_CONVOLUTION_FWD_ALGO_WINOGRAD_NONFUSED
+    };
+    size_t max_ws_size = getMaxWorkspaceSize<cudnnConvolutionFwdAlgo_t>(handle,conv,algo,sizeof(algo)/sizeof(algo[0]),state);
+    Workspace ws(state, max_ws_size);
+
+    CHECK(cudnnFindConvolutionForwardAlgorithmEx(handle, conv.idesc.desc, in,
+        conv.wdesc.desc, wght, conv.cdesc.desc, conv.odesc.desc, out, 1, &algoCount, 
+        &perfResults, ws.data, ws.size));
```
Comment (apaszke) [inline torch/csrc/cudnn/Conv.cpp]: I talked to @colesbury and we've decided that we should flush the block cache of the allocator after the algorithm is found (to get rid of any unnecessarily large blocks). You'd need to wrap this call and workspace in a separate block, and call `emptyCache` on the caching allocator. We don't have any API exposed for this purpose but it should be quite straightforward ([the relevant lines are here](https://github.com/aromnvidia/pytorch/blob/cb1dfa62c80aaa64689784fde1e7cbfc7eaf777f/torch/lib/THC/THCCachingAllocator.cpp#L456-L482)). It'd be best to add this function to `torch/lib/THC/THCCachingAllocator.h`.

**Review by apaszke — COMMENTED** [review]
The PR looks good now. We only need one more change so we don't make the allocator cache unnecessarily large.

File: torch/csrc/cudnn/Conv.cpp (hunk 4 above)
Comment (aromnvidia) [inline torch/csrc/cudnn/Conv.cpp]: Hi Adam,
I need a clarification on this.  I use Workspace class which deallocates memory chunk on exit [(Conv.cpp).](https://github.com/aromnvidia/pytorch/blob/cb1dfa62c80aaa64689784fde1e7cbfc7eaf777f/torch/csrc/cudnn/Conv.cpp#L94-L97)  Do you want to deallocate all unused memory?

**Comment by aromnvidia** [comment]
Sorry, Adam. Removed your comment by mistake. I thought I remove mine.

I thought it's not the right place to call emptyCache after `CHECK(cudnnFindConvolutionForwardAlgorithmEx(handle, conv.idesc.desc, in..`. We will fall into deadlock on `cuda_free_mutex`. I'd suggested to flush the cache in `chooseAlgorithm()` after `findAlgorithm(state, handle, conv, benchmark, in, out, wght, algo);`

**Comment by apaszke** [comment]
Yeah sure. Everything is ok as long as the cache is going to be flushed only when benchmarks are run (this should only happen a few times during training). 

Actually I've just started wondering if it wouldn't be simpler to limit the max workspace size to `max(largest_cached_free_block, free_device_memory)` and use use a call to `cudaMalloc` (without using THC) to allocate the benchmarks workspace. This shouldn't interfere with the cache at all, and wouldn't require the flush.

**Comment by ngimel** [comment]
You can't cudaMalloc largest_cached_free_block without first emptying cache, right? And in some unfortunate cases subsequent cudaFree'ing can interfere with nccl (though in some unfortunate cases that may happen also when caching allocator can't allocate requested block and starts cleaning up). Grrr, this is so hard to solve cleanly.

**Comment by apaszke** [comment]
Right, I meant to use the max and:
* if there's more free space on the device then use cudaMalloc,
* if the largest cached (and free!) block is larger, use that amount with the caching allocator. This shouldn't alter its state.

cudnnFind (no ex) used to call cudaFree too, and this part is already guarded by a free mutex.

**Comment by colesbury** [comment]
We discussed this earlier and I don't think these heuristics are a good idea. If you're worried that the caching allocator will unnecessarily hold on to large blocks, call `THCCachingAllocator_emptyCache` after benchmarking.

**Review by apaszke — APPROVED** [review]

**Comment by soumith** [comment]
@aromnvidia looks like you haven't tested the GPU builds after you pushed your latest commits? They actually fail with basic compiler errors.
https://build.pytorch.org/job/pytorch-PR-py2-linux/38//console

**Review by apaszke — APPROVED** [review]

**Comment by soumith** [comment]
this is now merged into master.

**Pull request closed without merging**